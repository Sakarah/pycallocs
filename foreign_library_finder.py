import allocs
import importlib
import importlib.abc
import importlib.machinery
import sys

# These can be modified by the user to change the finder search paths
lib_extension = ".so"

# The blank search path has a special meaning as dlopen is using its own search 
# paths when the file name does not contain any /
search_paths = [""]

class ForeignLibraryFinder(importlib.abc.MetaPathFinder):
    """
        Meta path finder searching foreign libraries usable
        with allocs.ForeignLibraryLoader.
    """

    def find_spec(fullname, path, target):
        if path is not None:
            # Should we support submodules ?
            return None
        
        for base_path in search_paths:
            try:
                filename = base_path + fullname + lib_extension
                loader = allocs.ForeignLibraryLoader(filename)
                return importlib.machinery.ModuleSpec(fullname, loader, origin=filename)
            except ImportError:
                continue

        return None

sys.meta_path.append(ForeignLibraryFinder)
