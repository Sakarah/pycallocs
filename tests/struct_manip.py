import foreign_library_finder
foreign_library_finder.search_paths.append("libs/")
import composite as m

a = m.hello_world()
print(a)
a.hello = 42
a.world = 9.99
print(a)

b = m.hello_world(1, 1)
print(b)

c = m.hello_world(hello = 2, world = float('nan'))
print(c)

d = m.hello_world(5)
print(d)

e = m.hello_world(d)
d.world = float('inf')
print(d)
print(e)
