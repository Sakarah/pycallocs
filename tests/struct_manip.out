(hello_world){hello: 0, world: 0.0}
(hello_world){hello: 42, world: 9.99}
(hello_world){hello: 1, world: 1.0}
(hello_world){hello: 2, world: nan}
(hello_world){hello: 5, world: 0.0}
(hello_world){hello: 5, world: inf}
(hello_world){hello: 5, world: 0.0}
